File: CRM-Client/src/schemas/product.ts
import { z } from 'zod'

export const BooleanFlagSchema = z.preprocess((value: unknown): unknown => {
  if (typeof value === 'boolean') return value
  if (value === 1 || value === '1') return true
  if (value === 0 || value === '0') return false
  return value
}, z.boolean())

export const ProductModuleRoleSchema = z.enum(['BASE', 'ADD_ON'])
export type ProductModuleRole = z.infer<typeof ProductModuleRoleSchema>

const PublicIdSchema = z.string().min(1)

export const ProductModuleResponseSchema = z.object({
  id: PublicIdSchema,
  public_id: PublicIdSchema,
  name: z.string(),
  description: z.string().nullable(),
  module_role: ProductModuleRoleSchema,
  is_active: BooleanFlagSchema,
  sort_order: z.number().int(),
  created_by: z.string(),
  updated_by: z.string().nullable(),
  created_time: z.string().min(1),
  updated_time: z.string().min(1),
})
export type ProductModuleResponse = z.infer<typeof ProductModuleResponseSchema>

export const ProductResponseSchema = z.object({
  id: PublicIdSchema,
  public_id: PublicIdSchema,
  name: z.string(),
  description: z.string().nullable(),
  is_active: BooleanFlagSchema,
  created_by: z.string(),
  updated_by: z.string().nullable(),
  created_time: z.string().min(1),
  updated_time: z.string().min(1),
  modules: z.array(ProductModuleResponseSchema),
})
export type ProductResponse = z.infer<typeof ProductResponseSchema>
export const ProductListResponseSchema = z.array(ProductResponseSchema)

const NameSchema = z.string().trim().min(1, '请输入名称').max(100, '名称不能超过100个字符')
const DescriptionSchema = z.string().max(2000, '描述不能超过2000个字符').nullable().optional()

export const ProductCreateSchema = z.object({ name: NameSchema, description: DescriptionSchema })
export type ProductCreate = z.infer<typeof ProductCreateSchema>
export const ProductUpdateSchema = z.object({ name: NameSchema.optional(), description: DescriptionSchema, is_active: z.boolean().optional() })
export type ProductUpdate = z.infer<typeof ProductUpdateSchema>

export const ProductModuleCreateSchema = z.object({
  name: NameSchema,
  description: DescriptionSchema,
  module_role: z.literal('ADD_ON').default('ADD_ON'),
  is_active: z.boolean().default(true),
  sort_order: z.number().int().min(0).default(0),
})
export type ProductModuleCreate = z.infer<typeof ProductModuleCreateSchema>
export const ProductModuleUpdateSchema = z.object({
  name: NameSchema.optional(),
  description: DescriptionSchema,
  is_active: z.boolean().optional(),
  sort_order: z.number().int().min(0).optional(),
})
export type ProductModuleUpdate = z.infer<typeof ProductModuleUpdateSchema>

File: CRM-Server/app/schemas/product.py
from __future__ import annotations

from datetime import datetime  # noqa: TC003  # Pydantic resolves this annotation at runtime.
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

ModuleRole = Literal["BASE", "ADD_ON"]

class ProductIntentRef(BaseModel):
    model_config = ConfigDict(extra="forbid")

    public_id: str
    name: str


class ProductModuleCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=2000)
    module_role: Literal["ADD_ON"] = "ADD_ON"
    is_active: bool = True
    sort_order: int = Field(0, ge=0)


class ProductModuleUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=2000)
    is_active: bool | None = None
    sort_order: int | None = Field(None, ge=0)


class ProductModuleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., validation_alias="public_id")
    public_id: str
    name: str
    description: str | None
    module_role: ModuleRole
    is_active: bool
    sort_order: int
    created_by: str
    updated_by: str | None
    created_time: datetime
    updated_time: datetime

    @field_validator("is_active", mode="before")
    @classmethod
    def coerce_active(cls, value: object) -> bool:
        return bool(int(value)) if isinstance(value, int) else bool(value)


class ProductCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=2000)


class ProductUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=2000)
    is_active: bool | None = None


class ProductResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., validation_alias="public_id")
    public_id: str
    name: str
    description: str | None
    is_active: bool
    created_by: str
    updated_by: str | None
    created_time: datetime
    updated_time: datetime
    modules: list[ProductModuleResponse] = Field(default_factory=list)

    @field_validator("is_active", mode="before")
    @classmethod
    def coerce_active(cls, value: object) -> bool:
        return bool(int(value)) if isinstance(value, int) else bool(value)
